### DISCUSSION TITLE
TallyArbiter port allocation

### DISCUSSION THREAD

**Discussion by syoffe** [josephdadams/TallyArbiter#516] [Q&A] [answered]
Hello,

First of all, I have just stumbled across this amazing project - thank you so much for releasing this open source!

In order to set up my tally light, I need to port forward the traffic. The server uses port 4455 but the connecting devices appear to use a "random" port. I was wondering if it is possible to set or at least restrict the port range? Basically, my tally light is on another subnet, so I need the device in the middle to pass the packets on. An IP route entry is not an option in this particular configuration.

Thanks for any advice,
Sam

**Comment by hrueger** [comment, marked as answer]
Hi, what do you mean by "connecting devices"?

**Reply by syoffe** [reply]
Hello, I mean the device connecting to the tally server, listening for tally data. For example, web browser client, M5, ...

**Reply by hrueger** [reply]
The web tally uses the same port as the web configuration interface, which is 4455. I believe (but don't quote me on that ;-) ) that the hardware listeners like the M5 use the same websocket api, therefore also on port 4455. Is that correct @MatteoGheza @JTF4 ?

**Reply by syoffe** [reply]
Looking at tcpdump, it seemed like the listener was sending/receiving from 51xxx, which imI guess is the browser websocket?

**Reply by syoffe** [reply]
Great if non-browser listeners use 4455, I don't have one to test with yet!

**Reply by JTF4** [reply]
> The web tally uses the same port as the web configuration interface, which is 4455. I believe (but don't quote me on that ;-) ) that the hardware listeners like the M5 use the same websocket api, therefore also on port 4455. Is that correct @MatteoGheza @JTF4 ?
> 
> 

That's correct. Hardware listener clients use 4455 for socketio. 

I'm not 100% sure what is happening in the 51xxx range though.

**Reply by syoffe** [reply]
Interesting, there was no traffic from or to 4455 on the listener when using the browser-based client (traffic was from/to 4455 on the server). Will try from an M5 when it arrives or a pi. Thanks a lot for the very quick help.

**Reply by syoffe** [reply]
Here is tcpdump listening to port 4455 on the server (172.20.10.2 is the server, 172.20.10.1 is the listener):
```19:19:59.425345 IP 172.20.10.2.4455 > 172.20.10.1.51108: Flags [P.], seq 369:536, ack 1, win 2048, options [nop,nop,TS val 3146528664 ecr 2862901317], length 167
19:19:59.425986 IP 172.20.10.2.4455 > 172.20.10.1.51108: Flags [P.], seq 536:738, ack 1, win 2048, options [nop,nop,TS val 3146528664 ecr 2862901317], length 202
19:19:59.435757 IP 172.20.10.1.51108 > 172.20.10.2.4455: Flags [.], ack 738, win 4096, options [nop,nop,TS val 2862902231 ecr 3146528664], length 0
19:20:00.429938 IP 172.20.10.2.4455 > 172.20.10.1.51108: Flags [P.], seq 738:905, ack 1, win 2048, options [nop,nop,TS val 3146529669 ecr 2862902231], length 167
19:20:00.430571 IP 172.20.10.2.4455 > 172.20.10.1.51108: Flags [P.], seq 905:1107, ack 1, win 2048, options [nop,nop,TS val 3146529669 ecr 2862902231], length 202
19:20:00.455098 IP 172.20.10.1.51108 > 172.20.10.2.4455: Flags [.], ack 1107, win 4096, options [nop,nop,TS val 2862903252 ecr 3146529669], length 0
19:20:01.433946 IP 172.20.10.2.4455 > 172.20.10.1.51108: Flags [P.], seq 1107:1274, ack 1, win 2048, options [nop,nop,TS val 3146530673 ecr 2862903252], length 167
19:20:01.434451 IP 172.20.10.2.4455 > 172.20.10.1.51108: Flags [P.], seq 1274:1476, ack 1, win 2048, options [nop,nop,TS val 3146530673 ecr 2862903252], length 202
19:20:01.538078 IP 172.20.10.1.51108 > 172.20.10.2.4455: Flags [.], ack 1476, win 4096, options [nop,nop,TS val 2862904335 ecr 3146530673], length 0```

**Reply by josephdadams** [reply]
all the clients use TCP for connection, so why does it matter if the listener client is on another subnet?

**Reply by JTF4** [reply]
Ahhhhh. I believe that you are running into one of the funny things about the TCP protocol. 

The server will communicate on port 4455, but when the client initiates contact, it doesn't necessarily use the same port. 

You don't have to forward the port that the listener client is connecting with, as it's not behind your firewall.

**Reply by syoffe** [reply]
Ah OK, please excuse what is probably a trivial question (I'm pretty new to networking). If the listener knows how to route to the server and initiates the connection (by registering as a listener), then does the server always know how to route to the listener? Even if the server doesn't have a route to the listener (i.e. it can't ping it)?

**Reply by syoffe** [reply]
Just to check my understanding:
`listener:51xxx |--- (subnet 1) ---> bridge device --- (subnet 2) ---> server:4455`
On the `bridge device` I do not need to port forward 51xxx to the listener? Packets from the server will know how to hop the bridge and route to the listener without my intervention?

**Reply by syoffe** [reply]
Just to say thank you - you are quite right, TCP witchcraft seems to take care of it!

**Reply by JTF4** [reply]
> Just to check my understanding:
> `listener:51xxx |--- (subnet 1) ---> bridge device --- (subnet 2) ---> server:4455`
> On the `bridge device` I do not need to port forward 51xxx to the listener? Packets from the server will know how to hop the bridge and route to the listener without my intervention?

Yes. This is correct. You don't need to port forward to the listener because the magic behind the tcp protocol handles it.
